### ISSUE TITLE
Is it possible to support GraphQL

### ISSUE THREAD

**Issue by darting** [Squidex/squidex#52]
hi there, just want to know if it is possible to support GraphQL?  (or do you have any plan to support it?)

thanks

**Comment by SebastianStehle** [comment]
Hi darting,

I think this should not be a problem. Squidex contains a schema object that exports the schema to JSON Schema and EDM and it should not be a big deal to create another schema for GraphQL. Then you can use a project like https://github.com/graphql-dotnet/graphql-dotnet to make the actual query to the database.

This is the other side of the story: How do you do efficient queries? In general there are several options:

1. Create indices manually
2. Use a database that just indexes everything (e.g. DocumentDB)
3. Use no indices at all (good solution for small datasets)
4. Use some autotuning (observe the queries and create indices for frequent patterns): Hard to implement.
5. Maintain indices by yourself (e.g. you could use GraphQL for materialized views).
6. Export the data to specialized tables (e.g. Elastic Search for Autocompletion and Fuzzy Search)

I am not sure yet which way to go. I would prefer not to have a vendor lock to database and this also limits the option.

**Comment by SebastianStehle** [comment]
@darting What are your expectations for GraphQL implementation? The definition of fields and structure is more or less defined by GraphQL but for the queries there is a lot of room for being "creative".

**Comment by SebastianStehle** [comment]
There is a readonly GraphQL API now :)

**Comment by zhangpeng-kooboo** [comment]
Hi @SebastianStehle ,

I was testing a simple GraphQL OData filter via the explorer.

```
{
  queryPerformanceContents (filter: "id eq '6c23dea5-22b5-44a5-9f94-65471ade0722'") {
    id,
    url
  }
}
```

Returns a error back.
```
{
  "data": {
    "queryPerformanceContents": null
  },
  "errors": [
    {
      "message": "Error trying to resolve queryPerformanceContents.",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ]
    }
  ]
}
```

Seem the GraphQL API does not work correctly?

**Comment by SebastianStehle** [comment]
The query looks incorrect. There is no field "url" directly. Should be data/url/iv. There is also a resolver/method to fetch an entity by id.

**Comment by zhangpeng-kooboo** [comment]
@SebastianStehle the url is there when I executed without a filter.

I tried the following query in the explorer, still the same error.
```
{
  queryPerformanceContents (filter: "data/name/iv eq 'Abdelhadi Baaddi'") {
    id,
    data{
    	name {
        iv
      }
    }
  }
}
```

**Comment by SebastianStehle** [comment]
I will have a look

**Comment by zhangpeng-kooboo** [comment]
Thks, it is a great product by the way. I was impressed.

**Comment by SebastianStehle** [comment]
Thank you. Btw: Please create a new issue the next tme.

**Comment by SebastianStehle** [comment]
For me it works:

```
query {
  queryPerformanceContentsContents(filter: "data/name/iv eq 't'") {
    id,
    data {
      name {
        iv
      }
    }
  }
}
```

**Comment by zhangpeng-kooboo** [comment]
Hi @SebastianStehle ,

Are there different versions between Hosted / Non-hosted.
I am using cloud.squidex.io.

Without a filter works fine.
![without filter](https://user-images.githubusercontent.com/4378651/38661471-4910279e-3e31-11e8-8780-3bee875300b5.png)

With a filter does not work.
![with filter](https://user-images.githubusercontent.com/4378651/38661470-48f57df4-3e31-11e8-8a72-ed8800917ced.png)

**Comment by zhangpeng-kooboo** [comment]
@SebastianStehle , 
Do you think is it something wrong in my account? It's a cloud version I could not see any details :(.
![with query](https://user-images.githubusercontent.com/4378651/38662303-25235204-3e34-11e8-85c3-4167205bde3d.png)

**Comment by SebastianStehle** [comment]
I don't know. Can you add me (sebastian@squidex.io) to your app?

**Comment by SebastianStehle** [comment]
Can you test the same query in the contents UI?

Just open the small triangle and add it to the filter input to generate the search term.

**Comment by zhangpeng-kooboo** [comment]
@SebastianStehle Just included you in my app.

**Comment by zhangpeng-kooboo** [comment]
@SebastianStehle right, just found out the field name is case sensitive. so data/Name/iv eq 'hello world' instead of data/name/iv eq 'hello world'

**Comment by SebastianStehle** [comment]
That's strange. Do you have a label for your field?

**Comment by zhangpeng-kooboo** [comment]
@SebastianStehle For the other fields under the same schema had labels, but not the above field.

**Comment by SebastianStehle** [comment]
Okay, but you have a Pascal case field name?

**Comment by zhangpeng-kooboo** [comment]
@SebastianStehle yes we do, e.g. BigImage, SmallImage and so on.
